### ISSUE TITLE
waitFor + getByRole causing severe delays

### ISSUE THREAD

**Issue by herecydev** [testing-library/dom-testing-library#820]
- `@testing-library/dom` version: 7.26.6
- `node` version: 14.3.0

### Relevant code or config:

```js
it("Causes a big delay", async () => {
      // This fires a fetch request
      userEvent.click(screen.getByRole("button", { name: "Search"}));

      await waitFor(() => {
        //super fast call
        //screen.getByText("some_link_text");

        // causes massive delays
        screen.getByRole("link", { name: "some_link" });
      });
    });
```

### What you did:

Running this test causes the fetch request promise to resolve a lot longer than usual (around 1.5 - 2s). I initially investigated this by wrapping `console.time` around the request.

![image](https://user-images.githubusercontent.com/11328618/99072037-27a9b200-25ab-11eb-817b-1d5772c0ecad.png)

I also had an additional test which looks very similar to the above, but instead of using `getByRole` it was using `getByText` to search for some text. That test took a fraction of the time:

![image](https://user-images.githubusercontent.com/11328618/99072114-52940600-25ab-11eb-8306-132e2cd3f471.png)

### What happened:

After changing `getByRole` to `getByText` **or** inserting a `getByText` query before the `getByRole` it resolved much faster. My working theory is that `getByRole` is so expensive that it was monopolising the `waitFor` loop and takes a long time to relinquish control to the pending promise that needs to resolve.

**Comment by nickmccurdy** [comment]
The issue is that `getByRole` is fairly slow, but it's also syncronous, so `waitFor` keeps rerunning it and has to wait for each call to finish before other code can run. It could potentially use some performance improvements, but perhaps we could make it asyncronous (or add an alternative for back compatibility) so this can be run asyncronously.

**Comment by herecydev** [comment]
Thanks for the quick reply Nick, does that not suggest that any `waitFor` that is sufficiently slow would trigger this? If so, is `waitFor` problematic in it's design?

What do you think the best short term resolution for this?

**Comment by nickmccurdy** [comment]
I'm guessing `waitFor` may be so eager to quickly rerun longer functions that it might actually be slowed down. You could try increasing the `interval` option, it defaults to only 50 milliseconds.

**Comment by herecydev** [comment]
For reference the `getByRole` was taking 400-500ms and would cause the `waitFor` to loop 3-4 times. The best fix (for time) is to insert a more trivial query before that (like a `getByText`) as that flushes that promise much faster.

Would you like me to document some of this? Feels like a sufficiently frustrating morning that I don't want others to go through!

**Comment by nickmccurdy** [comment]
A higher interval might cause less loops. Even if it doesn't make this specific test run shorter, a longer interval could allow other tests to run in parallel more effectively. Can you try this out?

Alternatively, we could dynamically adjust the interval by default based on how long it takes each time. I'm not sure if that's worth the effort though.

**Comment by herecydev** [comment]
Setting an interval of 750 still caused 2 iterations to fail (~450ms), then the fetch promise resolves. There's one more failure (output still not rendered to the screen), finally it passes.

Setting an interval underneath the `getByRole` time causes it to fail as expected. I.e. setting it to 200ms will cause 3 iterations and then the waitFor will just give up (total time is 1s by default if I remember).

I guess it works, but you keep having to up this arbitrary interval to "match" however long the waitFor will take as the complexity grows. That doesn't feel great to me.

**Comment by nickmccurdy** [comment]
Yea, I was thinking maybe `waitFor` could intelligently learn how long previous calls took and use that to adjust the interval. I'm not sure how well it would work in practice though.

**Comment by herecydev** [comment]
The problem is when the knowledge has been propogated differently;

https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-the-wrong-query
https://testing-library.com/docs/guide-which-query/#priority

There's absolutely nothing in the docs about the downsides of `ByRole`, and previously I've been happy to accept the small performance hit for the better testing methodology. But this issue is straight out problematic now.

I would love to get an action on how best proceed with this, do we:

1. auto-increment `waitFor` intelligently
2. Change inner workings of `ByRole` to be asynchronous
3. Provide documentation around this known issue with suitable guidance

etc.

**Comment by nickmccurdy** [comment]
1 would be cool to experiment with but as I said previously, though I'm not sure how difficult to implement and useful it would be, so it may not be worth it. I'm fine with either 2 or 3.

@kentcdodds @eps1lon What do you think about an asyncronous version of `*byRole` (2) to work around performance issues like this?

**Comment by kentcdodds** [comment]
One question to make sure I understand. What is your test waiting on that takes more than a single 50ms interval? Most/All tests using this utility should mock out any async code so it resolves on the next tick or two of the event loop 🤔

**Comment by kentcdodds** [comment]
Just noticed the screenshot, sorry about that. Looks like it's getting mocked by MSW, so I wouldn't expect this to take longer than a few milliseconds. Something is up there.

The `*ByRole` query is definitely the performance hog of the bunch, but it's a cost with the benefit of confidence which is why it's the top most recommended query. I suppose we could add a note to the docs warning about this. Probably a good idea. I'd love to make that query faster, but I still recommend it.

In this situation, there's something especially fishy going on. I think we'd need a reproduction to dive any deeper though.

**Comment by herecydev** [comment]
@kentcdodds yer to confirm MSW is registering is ~1ms so it's very fast. It's the `*ByRole` that's exceeding the `waitFor` interval window.

Reproduction will take some time, is it not clear that exceeding the `waitFor` interval window will cause this?

**Comment by eps1lon** [comment]
If you want to optimize tests for performance go through these steps until you're satisfied. They're ordered by their confidence impact (lower impact, higher ranked) then perf impact (higher impact, higher ranked).
1. Make sure you use the latest version of `jsdom`
1. don't use JSDOM but modern browsers
1. `getByRole(role, { hidden: true })` (WARNING: includes inaccessible elements)
1. `getByTestId` instead of `getByRole`
1. don't filter by `name` in `getByRole` (Can rely on order in `getAllByRole`. Order of elements in the DOM is in most cases not an implementation detail.)

Not using `jsdom` at all is the best solution since it has better perf and higher confidence.

**Comment by kentcdodds** [comment]
@eps1lon great walkthrough 👍 I think it makes sense to add this to the docs.

**Comment by eps1lon** [comment]
> though I think I'd prefer to try another query before testIDs. I think it makes sense to add this to the docs.

Definitely! It was easier to say since "use another query" might not be that helpful. If you're really blocked by perf problems I would just suggest a quick escape hatch and revisit it once the test breaks on unrelated changes (though that might be even harder to judge). It's just really hard to give actionable advise that people can apply to different scenarios.

I think the most generic advise is to re-evaluate your `waitFor` usage. If you're in a unit-test-like environment you should be able to mock the timers so that you can deterministically forward to the point where your assertion should be true (skipping the wasteful polling approach). In e2e tests (with real timers/requests) I'd like to see a reproduction since I would expect a browser to not have perf problems with `getByRole(role, { name })`.

And to give an optimistic outlook: The `ByRole` implementation is just a polyfill at the moment. Long term we'll use actual DOM APIs that are hopefully faster (see https://wicg.github.io/aom/spec/#computed-accessibility-tree).

**Comment by alexkrolick** [comment]
I would definitely try item 3 from @eps1lon's checklist for this particular case - the visibility calculation is expensive in large DOM trees
```
getByRole(role, { hidden: true })
```

**Comment by kentcdodds** [comment]
Thanks for the input here everyone. I think some docs here would go a long way.

@herecydev once you've tried and had success with one of these recommendations, would you mind working on improvement the documentation to help others who may face this in the future?

**Comment by eps1lon** [comment]
> I would definitely try item 3 from @eps1lon's checklist for this particular case - the visibility calculation is expensive in large DOM trees

To touch on another potential optimization point: In Material-UI we set the [default value](https://testing-library.com/docs/dom-testing-library/api-configuration#configuration) to `true` (fast option) in local development and to false (slow) in CI. The reason being that you rarely accidentally query for inaccessible elements. If you do, it should be obvious in the test which we'll check during code review (since an explicit `hidden` option stands out). In 1015 `ByRole` calls we only ever need to explicitly specify the `hidden` state in 25 cases. So in the vast majority of cases we just check the `hidden` state for confidence not because the test is actually about the `hidden` state.

The other part is that CI can be a little bit slower (since it already takes a couple of minutes). However, for local development you want snappy feedback so every millisecond counts.

That might be worth documenting. Though we shouldn't integrate that into the library since we don't know whether people have CI or if every commit passes through it.

**Comment by herecydev** [comment]
> 
> 
> If you want to optimize tests for performance go through these steps until you're satisfied. They're ordered by their confidence impact (lower impact, higher ranked) then perf impact (higher impact, higher ranked).
> 
>     1. Make sure you use the latest version of `jsdom`
> 
>     2. don't use JSDOM but modern browsers
> 
>     3. `getByRole(role, { hidden: true })` (WARNING: includes inaccessible elements - avoids expensive visibility check)
> 
>     4. `getByTestId` instead of `getByRole`
> 
>     5. don't filter by `name` in `getByRole` (Can rely on order in `getAllByRole`. Order of elements in the DOM is in most cases not an implementation detail.)
> 
> 
> Not using `jsdom` at all is the best solution since it has better perf and higher confidence.

Thanks Sebastian, to address these;

1. On the latest version
2. Can you elaborate on this one, we're using RTL with jest (and JSDOM). Is there a way to configure other environments, if so what would you recommend?
3. We're already setting this to true globally
4. I've already mentioned that I'm getting mixed messages with this one. @kentcdodds has clearly called this out as a common mistake: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#using-the-wrong-query
5. Similar to above, I want to look for a link that also has some text. Is this not preferable to combine into a single query

> I think the most generic advise is to re-evaluate your waitFor usage. If you're in a unit-test-like environment you should be able to mock the timers so that you can deterministically forward to the point where your assertion should be true (skipping the wasteful polling approach). In e2e tests (with real timers/requests) I'd like to see a reproduction since I would expect a browser to not have perf problems with getByRole(role, { name }).

I just want to stress this is nothing to do with timers. *Everything* is mocked. We're using MSW here, but the `*ByRole` query is taking longer than the `waitFor` and so waitFor times out.

**Comment by alexkrolick** [comment]
> Can you elaborate on this one, we're using RTL with jest (and JSDOM). Is there a way to configure other environments, if so what would you recommend?

ie Puppeteer, Cypress

How big is the DOM tree being queried in this test?

**Comment by meastes** [comment]
I was seeing similar performance issues in one of my tests as well. My app is using Material UI and I was seeing about 6 seconds per query with `getByRole()` making my total test time about 20 seconds. After switching the queries to `getByLabelText()` the test time went down to 600ms - HUGE difference. The component under test is a Drawer with about 50 checkboxes in it.

I agree that a small performance cost is worth the confidence, but it becomes unusable when that cost is about 6 seconds. I definitely agree in having some documentation updates to give visibility into this tradeoff.

**Comment by eps1lon** [comment]
> I just want to stress this is nothing to do with timers. Everything is mocked.

Why are you using `waitFor` if everything is mocked? If everything is mocked a single `jest.runAllTimers` should be faster than `waitFor`. Could you run `npx testing-library-envinfo react` and share the results?

> Can you elaborate on this one, we're using RTL with jest (and JSDOM). Is there a way to configure other environments, if so what would you recommend?

We're using mocha + karma with chrome headless and browserstack. Bundling the full test suite with webpack and running is as fast as running it with JSDOM. So there's a breakpoint where mocha + karma becomes faster if you get bundle time low enough. We don't have as severe problems with JSDOM so I suspect that this might be a real time saver for you.

However, I don't think it's appropriate for local development where you just run a subset of the tests.

> I was seeing about 6 seconds per query with getByRole()

6 full seconds for each call of `getByRole`? Could you run `npx testing-library-envinfo react` and share the results? If you could share a repro I might be able to understand the issue. Having a repro for these problems is important since that's the only way we can meaningfully work on the performance. Make sure you

**Comment by shan-du** [comment]
We are having similar performance issues with `getByRole`.
I did some comparison fetching the exact same element using `getByText` usually takes around 200~300ms, but with `getByRole` it takes upwards of 40 seconds!!! 

I know that both of these durations are a bit long anyways, but the point here is that relative to each other, `getByRole` is over 100 times slower, which is just ridiculous.

In the end I had to give the selectors a container scope like `container.getByRole` instead of `screen.getByRole`, then the amount of time reduces significantly, but in comparison, it is still a lot slower than `getByText`.

I'm really not sure what the reason is and perhaps it should really be mentioned in the documentations.

**Comment by herecydev** [comment]
@eps1lon sorry I can't be of much help now, this issue was quite painful and we move over to using cypress where this isn't a problem for us, as well as providing a suite of other benefits.

**Comment by eps1lon** [comment]
> I did some comparison fetching the exact same element using getByText usually takes around 200~300ms, but with getByRole it takes upwards of 40 seconds!!!

Please share a reproducible example. Otherwise we can't help you with this problem.

**Comment by alexkrolick** [comment]
I wonder if it has something to do with the MutationObserver optimizations in waitFor hogging too much of the thread for the more expensive getByRole DOM traversal

**Comment by meastes** [comment]
Here's a reproducible example. It's not quite as bad as my original project (about 5 seconds per getByRole call in the example instead of 6 seconds), but still shows that there's a huge performance issue. Hopefully this is helpful in debugging the issue!

https://github.com/meastes/get-by-role-example

**Comment by radfahrer** [comment]
I just ran into this and found that switching from `findByRole` with `name:` to `findByText` saved me over a second per test. I understand that `*byRole` is more specific, and thus more reliable for getting the correct element. However, as far as "confidence"  goes, I'm going to advise my team to avoid `findByRole`. The reason I found this performance issue is because I'm working on fixing a flaky test suite. The delay introduced by `findByRole` is causing our test suite to time out intermittently and therefore reduces confidence in our tests. 

I would highly recommend improving the performance of *byRole if possible.

**Comment by Duncan-Alexander-Coutts** [comment]
> Here's a reproducible example. It's not quite as bad as my original project (about 5 seconds per getByRole call in the example instead of 6 seconds), but still shows that there's a huge performance issue. Hopefully this is helpful in debugging the issue!
> 
> https://github.com/meastes/get-by-role-example
> 
> ```
>  PASS  src/components/DrawerExample.test.tsx (22.395 s)
>   DrawerExample
>     selecting
>       ✓ defaults checkboxes to checked when selectedColumns is specified (840 ms)
>       ✓ defaults checkboxes to checked when selectedColumns is specified (getByRole) (17729 ms)
> 
> Test Suites: 1 passed, 1 total
> Tests:       2 passed, 2 total
> Snapshots:   0 total
> Time:        23.846 s, estimated 30 s
> Ran all test suites matching /DrawerExample/i.
> ```

Apologies, not adding much to the conversation but my team has also been having problems with this. I can vouch that @meastes example is similar to what my team is seeing.

**Comment by mmnoo** [comment]
Just weighing in with something that worked for me. Its hacky, I dont fully understand _why_ it works, and I dont like it but it really sped up my test suite after the other things I tried didnt make a difference. 

My integration/e2e-ish tests that rendered my whole app were the slow ones causing problems, and especially slow were test in files with many other tests in the same file. I broke up those files to only have 1-3 tests (whatever it took to get the suite to run under 30 seconds), and that did the trick. I dont know why this works, but there seems to be an issue somewhere in the testing stack when multiple tests are in the same file.

**Comment by stabback** [comment]
I've spent the better part of a day taking a look at _why_ getByRole is much slower.  In our case, the slowdown is actually coming from `dom-accessibility-api`, currently used here:

https://github.com/testing-library/dom-testing-library/blob/84bf52f1b2b30f34e3a709e5bdd32e14623d6f60/src/queries/role.js#L150

If the textContent is pulled out of the node instead of using dom-accessibility-api to compute the accessible name, performance increases drastically.

Test time while using `computeAccessibleName`
> should show all keys in the layout (3665 ms)

Test time while comparing `element.textContent`
> should show all keys in the layout (300 ms)

In my test, that speeds my test up by 12 times!

Why?  I'm not too sure, but it may have something to do with jsdom's implementation of `getComputedStyle`, which historically has had poor performance.  [dom-accessibility-api](https://github.com/eps1lon/dom-accessibility-api)'s [computeAccessibleName](https://github.com/eps1lon/dom-accessibility-api/blob/2cf0211b8cd6ea2e7d6db0a677fde245fe0b2563/sources/accessible-name.ts#L32) function eventually calls `window.getComputedStyle`.  I found [an old bug](https://github.com/jsdom/jsdom/issues/1560) with jsdom which mentioned that they had poor performance and the ticket was closed after a bit of finger-pointing.  This was 5 years ago though, so it may not be relevant.

---

@kentcdodds - Would you support adding another option to the `*ByRole` selector which allows a user to specify that they are looking for text directly?  This would allow the selector to bypass `dom-accessibility-api` for instances where a user knows the name that they are looking for.  I'm happy to implement.

While a catch-all selector is nice to have, there are some serious performance gains to be had here while still maintaining confidence.

```ts
// Proposed use
getByRole('button', { text: 'Exact text match' });

// Proposed pseudo-code change
// https://github.com/testing-library/dom-testing-library/blob/84bf52f1b2b30f34e3a709e5bdd32e14623d6f60/src/queries/role.js#L150
    .filter(element => {
      if(options.text !== undefined) {
          return element.textContent === options.text
      }

      if (name === undefined) {
        // Don't care
        return true
      }

      return matches(
        computeAccessibleName(element, {
          computedStyleSupportsPseudoElements:
            getConfig().computedStyleSupportsPseudoElements,
        }),
        element,
        name,
        text => text,
      )
    })
```

**Comment by kentcdodds** [comment]
Thanks for the thorough analysis @stabback!

I could be convinced of this, but I don't know `dom-accessibility-api` at all (that's @eps1lon's domain) so maybe there's something else that could happen in there to reduce the compromises that need to made here. The issue is that you wouldn't be able to get a `textbox` with `text` because it doesn't have a `.textContent`, it has a label which has a text content. So adding a second way to specify the text you're looking for could lead to confusion.

I think that `hidden: true` (also globally configurable) as described above side-steps the computedStyle stuff which should speed things up and I think I'd prefer recommending that to the added confusion of an additional way to determine the text we're looking for...

**Comment by stabback** [comment]
The `hidden: true` bit does not impact my example at all, as the slowdown happens when calling `computeAccessibleName`.  The `hidden: true` bit is before that.  Toggling back and forth, and even deleting that code entirely from `@testing-library/dom` has a minimal impact (in my setup).

It'd be great if `dom-accessibility-api` could return an accessible name faster, but I believe the underlying problem is Jest's Window functions being slow.  I'm migrating this repository over from Karma, and this test runs lightning fast in Headless Chrome.  @eps1lon You'd know definitely know more here though!

Would updating the option from `text` to `textContent` help?  I don't want to propose adding an entire layer where we're matching arbitrary attributes, that seems contrary to the goals of testing-library.  Narrowing a role by the text content does seem like a common pattern though, I'd love to be able to speed that up.

I'm happy to get a PR together to help the discussion!

**Comment by kentcdodds** [comment]
I guess if you're ok with selecting by text rather than accessible name, then I suggest using the `*ByText` queries instead of `*ByRole` 🤷‍♂️

**Comment by stabback** [comment]
We definitely could switch all of our `*ByRole` queries to `*ByText` queries, but it'd be a shame to do that.  It would lower our confidence in our tests as the two do not query for the same thing, and we could have accidental a11y breaking changes (e.g. a junior dev switching a button to a div + onclick).

I do see the hesitancy to add any sort of non-end-user-focused options to such a user-focused query, but the performance hit here with jsdom is pretty substantial.  Faster tests encourage more tests!

I did see that there is a helper function to [get node text](https://github.com/testing-library/dom-testing-library/blob/84bf52f1b2b30f34e3a709e5bdd32e14623d6f60/src/get-node-text.ts#L3).  I'd change my recommendation to `getByRole('button', {nodeText: 'Hyperdrive!'})` to make it clear what is being selected.

**Comment by stabback** [comment]
Since my contribution to this thread is discussing a new feature, I've created a draft PR to move the conversation over and free up this thread for the original issue.

**Comment by eps1lon** [comment]
Just to be clear: If you have a problem switching to `ByNodeText` then a `nodeText` option has the exact  same drawback regarding robustness. I would argue that the `nodeText` approach is worse since it gives a false sense of security: "well I'm using `ByRole` which is robust, so the test is robust". 

There's a place for filtering by `textContent` (e.g. `listitem` causes oftentimes confusion since people want to query them by their content but they're not named by content but author). But that's a slightly different problem since we may also want to make sure people don't just query by content when the element is named by author+content. Though this is part of another discussion.

The accessible name as well as accessibility tree inclusion requires CSS. Calculating styles is slow in JSDOM which is why you see these drastic gains outside of JSDOM.

I was always hesitant to add an option to ignore styles since it's not trivial to teach how this affects `ByRole` (e.g. a11y tree inclusion changes unexpectly if you're not familiar with the spec which also affects what's part of the name).

It may make sense to add an option to `dom-accessibility-api` to ignore a11y tree inclusion (like `ByRole` does) and set that to the same value of `ByRole#hidden`. So if that's not already the case, that'd be a valued contribution that has to happen in `dom-accessibility-api` and `@testing-library/dom`.

**Comment by stabback** [comment]
> Just to be clear: If you have a problem switching to `ByText` then a `nodeText` option has the exact same drawback regarding robustness. I would argue that the `nodeText` approach is worse since it gives a false sense of security: "well I'm using ByRole which is robust, so the test is robust".

I get what you're saying, it's definitely less robust than filtering by the accessible name.  I'd love to eventually switch everything back over to the `name` filter if/when JSDOM narrows the gap in style calculation performance.  It's not a trivial gap. For context, I'm migrating a mid-size monorepo over with a few thousand tests from Karma to Jest.  Every package runs tests about twice as fast with Jest, but the packages with heavy use of `ByRole` run several times slower (~60s => ~300s).

> It may make sense to add an option to `dom-accessibility-api` to ignore a11y tree inclusion (like `ByRole` does) and set that to the same value of `ByRole#hidden`. So if that's not already the case, that'd be a valued contribution that has to happen in `dom-accessibility-api` and `@testing-library/dom`.

While I'm happy to do the work, this would be a breaking change for users of `@testing-library/dom`, wouldn't it?  Since `name !== nodeText`, then changing the behavior of the name filter when combined with `hidden` will affect some tests out there in the wild.  I do like the direction though and think that change may make sense within `dom-accessibility-api` regardless of what happens here.

**Comment by kentcdodds** [comment]
@stabback, I really appreciate that time you've taken to help improve these things. It's definitely one area of concern in this library. I like what @eps1lon suggested and if you're willing, would you like to open an issue on `dom-accessibility-api` to discuss this change there? I'm not concerned about pushing a major version to this package when necessary.

**Comment by stabback** [comment]
Happy to help if I can!  I'll get at it in the next few days.

**Comment by nstepien** [comment]
I've noticed that `waitFor` bypasses the set interval when mutations happen, for example with this
```js
test.only('test', async () => {
  render(<ComplexComponent />);

  let now = performance.now();

  await waitFor(
    () => {
      console.log(`check: ${performance.now() - now}ms`);
      now = performance.now();
      expect(1).toBe(2);
    },
    {
      interval: 2000,
      timeout: 6100
      // mutationObserverOptions: { subtree: false, attributes: true }
    }
  );

  console.log(`check: ${performance.now() - now}ms`);
});
```
I see this logged:
```
check: 1.0617000000002008ms
check: 54.92289999999957ms
check: 87.00720000000001ms
check: 24.19459999999981ms
check: 28.44000000000051ms
check: 37.776199999999335ms
check: 24.651099999999133ms
check: 1740.7797999999993ms
check: 1999.6796000000013ms
check: 1999.6800999999996ms
```

If I tweak the options with `mutationObserverOptions: { subtree: false, attributes: true }` then I get something more reasonable:
```
check: 1.095499999999447ms
check: 1999.2367999999997ms
check: 2001.2246999999998ms
check: 2000.5632999999998ms
```

So I think `waitFor`/`findBy` should respect the (default) interval, and throttle even when dom mutations happen. Otherwise it doesn't give enough room for async operations to run to completion. Especially when each `waitFor` run is longer than the set interval.

**Comment by nstepien** [comment]
Another benchmark:
```js
console.time('check');
let cancelButton = await screen.findByRole('button', { name: 'Cancel' });
console.timeEnd('check');
```
Result: `check: 2.586s`

```js
console.time('check');
let cancelButton;
await waitFor(async () => {
  await new Promise((resolve) => {
    setTimeout(resolve, 100);
  });
  cancelButton = screen.queryByRole('button', { name: 'Cancel' });
  expect(cancelButton).toBeInTheDocument();
});
console.timeEnd('check');
```
Result: `check: 1.096s`

So yeah, delaying things is ~2x faster 🤔

**Comment by alexkrolick** [comment]
Great investigation, @nstepien . Do you have a suggestion for what should change in the default implementation to unblock async operations?

**Comment by nstepien** [comment]
If `waitFor` had to respect the interval, then we wouldn't have a need for the mutation observer, which could make simpler cases worse, so maybe we shouldn't do that.

What if, instead, the mutation observer would schedule the `waitFor` run with `requestIdleCallback`?

```js
let idleCallbackId = null;

const mo = new MutationObserver(() => {
  if (idleCallbackId === null) return;
  idleCallbackId = requestIdleCallback(waitForCallback);
});

function nextInterval() {
  cancelIdleCallback(idleCallbackId);
  idleCallbackId = null;
  waitForCallback();
}
```

If there is always code running then it might never be truly "idle", but the interval checks will take care of that anyway.
This might actually work perfectly. Just needs some testing.

https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback
https://developer.mozilla.org/en-US/docs/Web/API/Window/cancelIdleCallback
It was also added to the TS types in the 4.4 release:
https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1029#issuecomment-869224737

I would also swap `setInterval` with a `setTimeout`, so there's an even interval between runs, otherwise we might end up with multiple runs back to back as they can take longer than the interval.

**Comment by nstepien** [comment]
`requestIdleCallback` isn't available in Node/JSDOM, gotta think of something different...

**Comment by alexkrolick** [comment]
Can we basically throttle the callback to some default like 50ms?

**Comment by nstepien** [comment]
Maybe we should just do
```js
const requestIdleCallback = window.requestIdleCallback ?? ((fn) => setTimeout(fn, 10));
```
So `requestIdleCallback` will be used when running in a browser with support, or otherwise fallback to a short `setTimeout`.
It should be good enough to relinquish the main thread to other async tasks.

**Comment by kentcdodds** [comment]
That seems like an acceptable solution. Is there a reason to have it scheduled for 10 ms? Could we not do this and achieve the same goal just faster?

```ts
const requestIdleCallback = window.requestIdleCallback ?? ((fn) => setTimeout(fn, 0));
```

**Comment by nstepien** [comment]
>Is there a reason to have it scheduled for 10 ms? Could we not do this and achieve the same goal just faster?

I'm not so sure, consider this very common case:
```js
const response = await fetch('api');
const json = await response.json();
```
the timeline would look something like this:
- `fetch()` is called in the main thread, it is awaited so the main thread is relinquished
- `setTimeout` immediately completes and queues its task
- `fetch()` resolves and queues its task
- the main thread is done
- the `setTimeout` task runs
- the `fetch()` task runs
- `response.json()` is called, main thread is relinquished again
- `setTimeout` immediately completes and queues its task
- `response.json()` resolves and queues its task
- the main thread is done
- the `setTimeout` task runs
- the `response.json()` task runs
- we finally get to render what we wanted

Some async operations depends on multiple tasks to run to completion to get what we want, but we run too much code in between.

Something similar actually happened to us with `react-focus-lock`, we had many `userEvent.tab()` before a `waitFor`, but `react-focus-lock` had many slow async tasks queued that we couldn't get past `response.json()` in time.

**Comment by kentcdodds** [comment]
Makes perfect sense 👍

**Comment by alexkrolick** [comment]
TL;DR the test code is blocking the app code because it's all on the same thread.

**Comment by nstepien** [comment]
I've opened #1031 to help address these issues.

**Comment by markerikson** [comment]
I can confirm that #1068 is a meaningful improvement. Pasting my comment from that thread:

I tried specifically adding a dep on `@testing-library/dom` to bump us to 8.11, and ran our longest UI test file, an "integration"-style test using `next-page-tester`.

Before:
```
 PASS  client/features/programDetails/characteristicsTab/tests/Characteristics.tests.tsx (107.118s)
  ProgramCharacteristicsContent
    √ Renders and allows editing Program Characteristics (98432ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        109.2s, estimated 124s
```

After:

```
 PASS  client/features/programDetails/characteristicsTab/tests/Characteristics.tests.tsx (90.958s)
  ProgramCharacteristicsContent
    √ Renders and allows editing Program Characteristics (82187ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        93.06s, estimated 106s
```

No other code changes.

So, this distinctly made _some_ improvement.

(I still don't think the test file should be taking anywhere this long to begin with, but I haven't had time to really drill down and see what else is taking it so long.)

**Comment by ahayes91** [comment]
It looks like this has been open for quite a while and has been tricky to resolve 😅 I usually recommend avoiding byRole inside waitFor syntax these days. Would there be any appetite for a linting rule in the meantime (something like https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/no-wait-for-side-effects.md) to prevent folks from running into this one? 👀

**Comment by MatanBobi** [comment]
Hi @ahayes91!
I don't think this is something we'll want to prevent (or advocate people to avoid). Using `*ByRole` is one of the best practices to see that our elements have the accessible role we need to make the app accessible.
If we do have performance issues in `*ByRole` we should tackle them :)
Thanks for the offer!

**Comment by berTrindade** [comment]
Hey guys, sorry for chiming in, I've been trying to understand where to use *finBy and waitFor. They seem to do pretty much the same, I know they're not equals. Could someone point me to a main difference that could help making this choice easier ?

**Comment by berTrindade** [comment]
Hi @MatanBobi, yeah! I understood that part. But I'm facing a scenario where I have a button which is disabled while a HTTP call is made and when I try to use findByRole it says it runs out of time, but when I use waitFor + getByRole it works for my assertion. I wonder why.

**Comment by MatanBobi** [comment]
@berTrindade I'd love to answer this in details but I don't think it should be a part of this issue.
Can you please either open a stackoverflow question and link to it here or reach out on our [Discord server](https://discord.gg/testing-library) or my twitter account and I'd love help :)

**Comment by berTrindade** [comment]
@MatanBobi true. You're absolutely right. Thanks and sorry 😅

**Comment by ghostd** [comment]
@MatanBobi i also ran into some timeout issues; i followed your advice and it fixed my tests: 

> I usually use `findBy*` for elements that should appear (even in a slight delay) and `waitFor` to wait for anything else (loader, assertion etc).

I'm not sure to understand why it's make a difference, maybe it's worth a paragraph in the documention/FAQ/whatever? What do you think?

**Comment by tkrotoff** [comment]
Did somebody tried with [Happy DOM](https://github.com/capricorn86/happy-dom) instead of [jsdom](https://github.com/jsdom/jsdom) to see if it helps?

**Comment by MatanBobi** [comment]
> @MatanBobi i also ran into some timeout issues; i followed your advice and it fixed my tests:
> 
> > I usually use `findBy*` for elements that should appear (even in a slight delay) and `waitFor` to wait for anything else (loader, assertion etc).
> 
> I'm not sure to understand why it's make a difference, maybe it's worth a paragraph in the documention/FAQ/whatever? What do you think?

Not sure what specifically helped you with my comment so if you'll explain that I'll be able to understand if we should emphasize something in our docs :)

**Comment by tkrotoff** [comment]
> Did somebody tried with [Happy DOM](https://github.com/capricorn86/happy-dom) instead of [jsdom](https://github.com/jsdom/jsdom) to see if it helps?

I tried: https://github.com/tkrotoff/getByRolePerfIssue

- With [jsdom](https://github.com/jsdom/jsdom):

```
 PASS  src/Form.test.tsx
  ✓ getByText() (473 ms)
  ✓ getByRole() (1454 ms)
```

- With [Happy DOM](https://github.com/capricorn86/happy-dom):

```
 PASS  src/Form.test.tsx
  ✓ getByText() (161 ms)
  ✓ getByRole() (150 ms)
```

**Comment by ghostd** [comment]
@MatanBobi what helped me is "replace the `await findBy` with `waitFor`"... but i did not understand why it works better this way.

**Comment by frankandrobot** [comment]
It's been 3 years and we're still saying the same perf issues even when using all the latest versions of the react testing libraries (as well as latest jsdom).

So had an idea for a workaround. Here's some psuedo code

```
const simpleGetByRole = (role, matcher) => {
    const element = screen.getByText(matcher)
    // assert element is the right role
```

Basically, it's getByText but we assert that the matched element is of the right type so... we don't lose some of the test confidence when using only getByText

However .... am wondering what edge cases this simple workaround can miss.

**Comment by MatanBobi** [comment]
@frankandrobot The problem with `getByRole` is that it does a lot of things and it's not `getByRole` that's slow, it's JSDOM in this case (4ms for `querySelectorAll` just as an example). The workaround you're suggesting here has some flaws because:
1. It doesn't account for implicit roles which is widely used.
2. It doesn't care about visibility.
3. The text of an element is not it's accessible name in all cases - it can also be implied from aria semantics.

We know that this is a pain point when using testing-library, I can only recommend giving happy-dom a try because it's faster than JSDOM.

**Comment by frankandrobot** [comment]
> It doesn't account for implicit roles which is widely used.

For my org that's likely the deal breaker. If we can make that work, this simpler version should be good enough. (2 can be done easily with a visibility check)

Yes, we've also tried happy-dom. In the past it just hasn't been ready for prime time.  

However, this morning, I just tried it on some complex tests that previously did not work with happy-dom. And i'm happy to report that all tests passed with the exception of an act warning

So i think definitely, we're ready to start re-evaluating happy-dom again

**Comment by MatanBobi** [comment]
`getByRole` and `getByText` have entirely different meanings and different purposes so comparing between them is an unfair fight, as I mentioned above, `ByRole` has way more things to do. Unfortunately, I do not have the capacity at the moment to try and think of a way to improve the performance here, but it's merely because we have a lot of DOM work to do.
Visibility check might sound easy, but it requires calling `getComputedStyle`, which calculates the stylesheet for the DOM element, and for **every** parent of the element because visibility is inherited. There's **a lot** to think about here. I'm not saying that there's no way to improve it, I'm just saying that it's not an easy task as one might think :)

**Comment by cmacdonnacha** [comment]
Looking for some opinions. Is the `byRole` query as vital when using a UI library like MUI, who pride themselves on having fully accessible components? We're really starting to see significant performance issues within our tests and I have always promoted the use of `byRole` to ensure our code is accessibility friendly. However, if we can "trust" that these checks are being done for the most part by MUI, we may look into  using `byText` more regularly in tests where performance is hit the most.

**Comment by mmnoo** [comment]
I've already started replacing `byRole`. It costs too much. `byText` isnt the same, yes, but it will have to do. Balancing tradeoffs.

**Comment by MatanBobi** [comment]
@cmacdonnacha, @mmnoo I gave a talk to explain what's happening in `ByRole` query to understand why it has a purpose. I recommend giving it a view (though I'm a bit biased):
https://www.youtube.com/watch?v=jNAMdsbdvlIs&ab_channel=ReactNext

**Comment by MatanBobi** [comment]
@frankandrobot, you're not, and we're 100% open for criticism.
I'm not saying it is what it is, we are trying to workaround those, having said that, the major bottleneck here is usually JSDOM.
Regarding your specific question:
1. If you don't care about visibility check for example, you can use [`hidden`](https://testing-library.com/docs/queries/byrole#hidden) for example to avoid that check and get a tiny speedup. The rest is probably too related to the role selection and we can't remove it.
2. I think that not using `ByRole` misses a bit of the accessibility checks and misses the point of our guiding principle.
3. I highly recommend trying `vitest` instead of `jest` and also `happy-dom` instead of `jsdom` for some quick wins :) 
Feel free to ask more questions, I'll do my best to help.

**Comment by frankandrobot** [comment]
We've used vitetest before

While it does help test boot up, if the problem is Jsdom, it doesn't provide speed ups to slow queries.

As for happy Dom, it's too incomplete to use in production. I really wish you guys would stop recommending it 😜


On Nov 6, 2023 11:01 AM, Matan Borenkraout ***@***.***> wrote:

@frankandrobot<https://github.com/frankandrobot>, you're not, and we're 100% open for criticism.
I'm not saying it is what it is, we are trying to workaround those, having said that, the major bottleneck here is usually JSDOM.
Regarding your specific question:

  1.  If you don't care about visibility check for example, you can use hidden<https://testing-library.com/docs/queries/byrole#hidden> for example to avoid that check and get a tiny speedup. The rest is probably too related to the role selection and we can't remove it.
  2.  I think that not using ByRole misses a bit of the accessibility checks and misses the point of our guiding principle.
  3.  I highly recommend trying vitest instead of jest and also happy-dom instead of jsdom for some quick wins :)
Feel free to ask more questions, I'll do my best to help.

**Comment by MatanBobi** [comment]
I have used `vitest` and `happy-dom` for production use as a replacement to `jest` + `jsdom` or else I wouldn't have recommended it. I can understand why **some** cases are still not supported, but you can definitely migrate some tests and run them side by side.

**Comment by frankandrobot** [comment]
Yes. I'm not saying it's not usable. But it's by no means a complete replacement.

To give an example, in my code base, in about a dozen test files. We're only able to migrate two to happy Dom.

That's too low to be a useful replacement.



On Nov 6, 2023 11:10 AM, Matan Borenkraout ***@***.***> wrote:

I have used vitest and happy-dom for production use as a replacement to jest + jsdom or else I wouldn't have recommended it. I can understand why some cases are still not supported, but you can definitely migrate some tests and run them side by side.

**Comment by sibelius** [comment]
what is the workaround ?

**Comment by sibelius** [comment]
how to migrate from jsdom to a real browser ?

we are using jest + testing library 

is there an environment for this ?

**Comment by eps1lon** [comment]
> how to migrate from jsdom to a real browser ?

This is out of scope for this discussion. Migrating from Jest+JSDOM to Playwright is not trivial. I'd suggest looking up Playwright's Getting Started docs to get a feeling how testing works in Playwright vs. Jest. In Playwright you wouldn't need `@testing-library/react` since the queries are built-into Playwright.

**Comment by sibelius** [comment]
is there a faster jsdom ?

**Comment by melan0802** [comment]
@eps1lon Have you considered replace `filter` with `find` in `queryAllByRole` when we use `getByRole`?
https://github.com/testing-library/dom-testing-library/blob/56543d51a8776d3bc7fa9b947771ce5308bf5df4/src/queries/role.ts#L266-L281

It will allow us to skip some expensive `computeAccessibleName` after we find our tartget.

**Comment by foxaltus** [comment]
Hey! I want to share my experience here as I've been struggling with a few slow tests that were consistently timing out.

In my case, the slow tests were the result of using `screen.findByRole("cell", { name: 'xxx' })` on tables (with less than 100 cells).

I tried migrating to `happy-dom`, which was pretty much effortless (minus a few issues caused by poorly written tests on my side). However, I only obtained marginal gains with `happy-dom`. Tests that were timing out with `jest-dom` were also timing out with `happy-dom`.

I tried using the [`hidden` option](https://testing-library.com/docs/queries/byrole#hidden) too (or the [global `defaultHidden`](https://testing-library.com/docs/dom-testing-library/api-configuration#defaulthidden)) of `testing-library` with marginal gains again, but not enough to solve the timeout.

The only thing that worked for me was to replace `findByRole` with `findByText`. Not ideal, but it makes a huge difference.

I am not sure if there is a particular issue with tables, but I would advise to stay away from accessible queries when working with them...

**Comment by TomPridham** [comment]
> @eps1lon Have you considered replace `filter` with `find` in `queryAllByRole` when we use `getByRole`?
> 
> https://github.com/testing-library/dom-testing-library/blob/56543d51a8776d3bc7fa9b947771ce5308bf5df4/src/queries/role.ts#L266-L281
> 
> It will allow us to skip some expensive `computeAccessibleName` after we find our tartget.

i don't think we want to do that because then the single element matchers wouldn't throw if there was more than one element

**Comment by frankandrobot** [comment]
We've been exploring a possible workaround. Here's some context:

* An element's inferred name is the inferred aria label. Ex: `<button>A button</button>` has an inferred aria label of "a button". There are also inferred roles. The button above has a role of "button", but almost anything can have a button role by just assigning role attribute a value of "button". 
* getByRole has quicker cousins, namely getByText and getByAriaLabel. (However, by themselves they don't check for role.)
* The workaround is to use one of these cousins to search by label or inferred name *and then* use `within` on the parent element to assert the role. In testing, this has been almost as fast as plain getByText!

But of course there are caveats:

1. Users need to know whether or not the aria-label is inferred. 
2. The inferred name can only be in the top-level of the component. Otherwise, selection will fail. For example, in `<td><div>a cell</div></td>`, the inferred name of the cell is "a cell", so the workaround fails---it gets the DIV and then checks whether or not it's a cell, so it fails.
3. The role check is done by calling `within#getAllByRole` (w/ no name) on the parent element. In edge cases, it's DOM tree can be just as complex as the root DOM tree, so there will be no speed up. (It may be possible to improve on this by temporarily detaching the node from the DOM.) 

Notes:

1. Yes, you do need to know implementation details but IMO it isn't a big of a deal. It's a bigger issue w/ third party libraries but IMO there should be test selector helpers for these anyway.
2. At least in our case, we use a shared component library, so instances of caveat 1) are limited to a handful of cases. 

The code:

```typescript
type GetByRoleFullOpts = Parameters<typeof original.getByRole>;
type GetByRoleOpts = GetByRoleFullOpts[1] & { nameType?: "inferred" | "aria-label" }

export const getByRole = (role: GetByRoleFullOpts[0], opts: GetByRoleOpts): GetByRoleReturn => {
  const { name, nameType, ...restOpts } = opts ?? {};

  if (name && typeof name !== "function") {
    const getter = nameType === "inferred" ? original.getByText : original.getByLabelText;
    const foundElement = getter(name, restOpts);
    const elements = within(foundElement.parentElement ?? document.documentElement).getAllByRole(role, restOpts);
    const match = elements.find((elem) => elem === foundElement);
    expect(match).toBeDefined();
    return foundElement;
  } else {
  return original.getByRole(role, opts);
};
```

**Comment by benlesh** [comment]
I've profiled issues with this in our codebase locally, and 90% of the speed issues are coming from checking to see if the element's subtree is "accessible", meaning if it's potentially visible to the user. The problem is it recursively checks every element, and worse, runs a `getComputedStyle` on it to see if its visible.

There's a new API for that now [checkVisibility](https://developer.mozilla.org/en-US/docs/Web/API/Element/checkVisibility), but that's probably not going to help much in node.

It's probably prudent to add an escape hatch here to skip this check. something like `getAllByRole('option', { strictVisibilityCheck: false })`

Honestly, I think even having this check to begin with is a bit pedantic given the nature of these tests. If it was my project, I would default to not even checking that unless someone had it on `getAllByRole('option', { strictVisibilityCheck: true })` or the like.

**Comment by MatanBobi** [comment]
Thanks @benlesh for the investigation.
This is already an option as stated by @eps1lon in [this comment](https://github.com/testing-library/dom-testing-library/issues/820#issuecomment-726936225) by using `getByRole('option', { hidden: true })`, which will skip the `isInaccessible` checks. This can also be turned off by default by calling `configure({defaultHidden: true})`, more info in our [docs](https://testing-library.com/docs/dom-testing-library/api-configuration#defaulthidden).

**Comment by hwride** [comment]
Last I checked `hidden: true` was not a particularly great improvement for performance of `getByRole`. So even if it does skip the accessibility checks, it doesn't seem to be the whole story/solution.

**Comment by benlesh** [comment]
For record keeping, I did submit a PR: https://github.com/testing-library/dom-testing-library/pull/1331


I'll explore other options for our needs. These tests are honestly slower than just running Playwright or Cypress tests at this point.

**Comment by benlesh** [comment]
@hwride you might want to check again. `hidden: true` improved performance of `getAllByRole` at least 10x in my problem test.

**Comment by eps1lon** [comment]
> I'll explore other options for our needs. These tests are honestly slower than just running Playwright or Cypress tests at this point.

That is what we recommend for high confidence testing. For lower confidence testing, disabling the inaccessible check via `hidden: true` is recommended for local testing and then enabling this check in CI.

**Comment by foxaltus** [comment]
> [@benlesh](https://github.com/benlesh) problem with `hidden: true` is that it does nothing for `description` or `name` options. If we have `name`, the code calls [computeAccessibleName](https://github.com/testing-library/dom-testing-library/blob/main/src/queries/role.ts#L273) which eventually calls [isHidden](https://github.com/eps1lon/dom-accessibility-api/blob/main/sources/accessible-name-and-description.ts#L71) which gets computed styles anyway and causes the slowdown. So while simple `*ByRole(role, { hidden: true })` gets speed up, using `*ByRole(role, { description, hidden: true })` or `*ByRole(role, { name, hidden: true })` is as slow as without `hidden`.

hey @Malgalad that's good insights and it explains why my *ByRole + name requests are still as slow as ever...

All hopes are on https://github.com/eps1lon/dom-accessibility-api/issues/1011 then. I've tried running testing-library with their latest release 0.7.1 (using pnpm overrides) as it is supposed to include the first performance enhancement, but it didn't make any difference.

**Comment by foxaltus** [comment]
Going down the rabbit hole... so *ByRole + name queries are slow because `getComputedStyles` is still getting called under the hood.

In `happy-dom`, which I used, the method is notably slow, as reported in https://github.com/capricorn86/happy-dom/issues/599. One of the workarounds given is:

> Another solution is to override getComputedStyle() with window.getComputedStyle = () => new CSSStyleDeclaration();

Although this did not work as such (Illegal constructor error), adding the following to my tests did work and cut the execution time by a factor of 10!

```ts  
const dummyStyle = document.createElement('span').style;
window.getComputedStyle = () => dummyStyle;
```

It's not pretty, but that would be a fair workaround for a slow test (mine would take 10 secs... which got reduced to 1sec).

**Comment by scttcper** [comment]
[latest jsdom 30.1 ](https://github.com/jsdom/jsdom/releases/tag/v30.1.0)has a bunch of fixes around performance in this area and i've opened https://github.com/testing-library/dom-testing-library/pull/1391 to speed it up slightly more across all dom implementations.

css visibility still probably pretty slow to check.
